asllib/Interpreter.ml

```diff
@@ -243,110 +297,153 @@ module Make (B : Backend.S) (C : Config) = struct
+            return (v, env) |: Rule.EGlobalVar
         | None -> (
             match IMap.find_opt x env.local.storage with
             | Some v ->
                 let* () = B.on_read_identifier x env.local.scope v in
-                return v |: Rule.ELocalVar
+                return (v, env) |: Rule.ELocalVar
             | None -> fatal_from e @@ Error.UndefinedIdentifier x))
     | E_Binop (op, e1, e2) ->
-        let* v1 = eval_expr env e1 and* v2 = eval_expr env e2 in
-        B.binop op v1 v2 |: Rule.Binop
+        let* v1, env = eval_expr env e1 in
+        let* v2, env = eval_expr env e2 in
+        let* v = B.binop op v1 v2 in
+        return (v, env) |: Rule.Binop
     | E_Unop (op, e) ->
-        let* v = eval_expr env e in
-        B.unop op v |: Rule.Unop
+        let* v, env = eval_expr env e in
+        let* v = B.unop op v |: Rule.Unop in
+        return (v, env)
     | E_Cond (e1, e2, e3) ->
-        B.bind_ctrl (eval_expr env e1) (fun v ->
-            B.ternary v
-              (fun () -> eval_expr env e2)
-              (fun () -> eval_expr env e3))
-        |: Rule.ECond
+        let* v =
+          B.bind_ctrl (eval_expr env e1) (fun (v, env) ->
+              B.ternary v
+                (fun () -> eval_expr_sef env e2)
+                (fun () -> eval_expr_sef env e3))
+        in
+        return (v, env) |: Rule.ECond
     | E_Slice (e', slices) ->
-        let* positions = eval_slices env slices and* v = eval_expr env e' in
-        B.read_from_bitvector positions v |: Rule.ESlice
+        let* v, env = eval_expr env e' in
+        let* positions = eval_slices env slices in
+        let* v' = B.read_from_bitvector positions v in
+        return (v', env) |: Rule.ESlice
     | E_Call (name, args, named_args) ->
-        let vargs = List.map (eval_expr env) args
+        (* TODO: rework to allow side-effects there. *)
+        let vargs = List.map (eval_expr_sef env) args
         and nargs =
-          let one_narg (x, e) = (x, eval_expr env e) in
+          let one_narg (x, e) = (x, eval_expr_sef env e) in
           List.map one_narg named_args
         in
-        let*| returned = eval_func env.global name (to_pos e) vargs nargs in
-        one_return_value e name returned |: Rule.ECall
+        let* returned, global =
+          eval_func env.global name (to_pos e) vargs nargs
+        in
+        let v = one_return_value e name returned in
+        return (v, { env with global }) |: Rule.ECall
     | E_Record (_, li, ta) ->
-        let one_field (x, e) = eval_expr env e >=> pair x in
-        let* fields = prod_map one_field li in
-        make_record e (type_of_ta e ta) fields |: Rule.ERecord
+        (* TODO: rework to allow side-effects in record declarations. *)
+        let one_field (x, e) = eval_expr_sef env e >=> pair x in
+        let* v =
+          let* fields = prod_map one_field li in
+          make_record e (type_of_ta e ta) fields |: Rule.ERecord
+        in
+        return (v, env)
```

Comment:
Hi @maranget, I should have added that in the PR description.

Indeed this PR does not fully implement side-effects. Here I have a problem because I want to use `ternary` which does not support side-effects.

Reply:
Agreed. In fact one could use `ternary` for side-effect-free expression, or even approximation of those. A resonable approximation would be expression with no function calls inside.

Reply:
That could work, I'll do that once the mem/pointer thing is decided.